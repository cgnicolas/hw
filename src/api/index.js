const url = 'http://localhost:3001';

const API = {
  call(url, method, body = {}) {
    const data = {
      method,
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
    };
    if (Object.keys(body).length > 0) {
      data.body = JSON.stringify(body);
    }
    return fetch(url, data).then((response) => {
      return response.json();
    });
  },

  get() {
    const path = '/getComments';
    return this.call(url + path, 'get');
  },

  post(body = {}) {
    const path = '/createComment';
    return this.call(url + path, 'post', body);
  },

  delete() {
    const path = '/deleteComments';
    return this.call(url + path, 'delete');
  },
};

export default API;
