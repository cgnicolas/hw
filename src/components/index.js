export { default as Form } from './Form';
export { default as CommentsList } from './CommentsList';
