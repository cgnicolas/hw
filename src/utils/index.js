export const sortComments = (comments) => {
    return comments.sort((a, b) => {
        const { created: aTimestamp } = a;
        const { created: bTimestamp } = b;
        const aDate = new Date(aTimestamp);
        const bDate = new Date(bTimestamp);
        return bDate.getTime() - aDate.getTime();
      });
}