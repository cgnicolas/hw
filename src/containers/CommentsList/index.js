import React from 'react';
import { CommentsList } from '../../components';

export default function CommentsListContainer({ comments = [] }) {
  return (
    <CommentsList>
      {comments.map((comment) => {
        return <CommentsList.Comment {...comment} key={comment.id} />;
      })}
    </CommentsList>
  );
}
