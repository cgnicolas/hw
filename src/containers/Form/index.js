import React, { useState } from 'react';
import { Form } from '../../components';
import { v4 as uuidv4 } from 'uuid';
export default function FormContainer({ onSubmit, loading }) {
  const [name, setName] = useState('');
  const [message, setMessage] = useState('');
  const formSubmission = (e) => {
    e.preventDefault();
    onSubmit({
      message,
      name,
      id: uuidv4(),
    }).then(() => {
      setName('');
      setMessage('');
    });
  };

  return (
    <Form onSubmit={formSubmission}>
      <Form.Input
        name="Name"
        placeholder="Enter your name"
        value={name}
        id="user-name"
        onChange={({ target: { value } }) => setName(value)}
      />
      <Form.CommentTextArea
        placeholder="Join the discussion!"
        value={message}
        id="user-message"
        onChange={({ target: { value } }) => setMessage(value)}
      />
      <Form.Button disabled={loading}>Comment</Form.Button>
    </Form>
  );
}
