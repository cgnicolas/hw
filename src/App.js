import React, { useEffect, useState } from 'react';
import { CommentsList, Form } from './containers/';
import API from './api';

function App() {
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);
  const [comments, setComments] = useState([]);
  /*
    Fetch Existing Comments
  */
  useEffect(() => {
    API.get()
      .then((response) => {
        setComments(response);
      })
      .catch((err) => {
        setError(err);
      })
      .finally(() => {
        setLoading(false);
      });
  }, []);

  const handleCreate = (commentBody) => {
    setLoading(true);
    return API.post(commentBody)
      .then(() => {
        return API.get();
      })
      .then((response) => {
        setComments(response);
      })
      .catch((err) => {
        setError(err);
      })
      .finally(() => {
        setLoading(false);
      });
  };

  return (
    <div className="appContainer">
      <Form loading={loading} onSubmit={handleCreate} error={error} />
      <CommentsList comments={comments} loading={loading} error={error} />
    </div>
  );
}

export default App;
