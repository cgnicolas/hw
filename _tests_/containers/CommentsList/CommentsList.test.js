import React from 'react';
import { mount } from 'enzyme';
import { CommentsList } from '../../../src/containers';

describe('<CommentsList/>', () => {
  const mockProps = {
    comments: [
      {
        id: 1,
        name: 'Cedric',
        created: 'a day ago',
        message: 'I give this Test Suite a 10 out of 10!',
      },
    ],
  };

  it('should render', () => {
    const renderedModule = mount(<CommentsList {...mockProps} />);
    expect(renderedModule).toMatchSnapshot();
  });

  it('should render with no comments provided', () => {
    const renderedModule = mount(<CommentsList />);
    expect(renderedModule).toMatchSnapshot();
  });
});
