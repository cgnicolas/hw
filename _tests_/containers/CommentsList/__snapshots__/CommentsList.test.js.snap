// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CommentsList/> should render 1`] = `
<CommentsListContainer
  comments={
    Array [
      Object {
        "created": "a day ago",
        "id": 1,
        "message": "Another great test case! Love to see it.",
        "name": "Cedric",
      },
    ]
  }
>
  <CommentsList>
    <styled.ul>
      <ul
        className="sc-eCApnc kacMPJ"
      >
        <Component
          created="a day ago"
          id={1}
          key="1"
          message="Another great test case! Love to see it."
          name="Cedric"
        >
          <styled.li>
            <li
              className="sc-jSFjdj fuuqgA"
            >
              <styled.p>
                <p
                  className="sc-gKAaRy"
                >
                  Another great test case! Love to see it.
                </p>
              </styled.p>
              <styled.small>
                <small
                  className="sc-iCoGMd"
                >
                  Cedric at a day ago
                </small>
              </styled.small>
            </li>
          </styled.li>
        </Component>
      </ul>
    </styled.ul>
  </CommentsList>
</CommentsListContainer>
`;

exports[`<CommentsList/> should render with no comments provided 1`] = `
<CommentsListContainer>
  <CommentsList>
    <styled.ul>
      <ul
        className="sc-eCApnc kacMPJ"
      />
    </styled.ul>
  </CommentsList>
</CommentsListContainer>
`;
